lib/ansible/modules/cloud/docker/docker_plugin.py

```diff
@@ -0,0 +1,267 @@
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = u'''
+module: docker_plugin
+version_added: "2.8"
+short_description: Manage Docker plugins
+description:
+  - Install/remove Docker plugins.
+  - Performs largely the same function as the "docker plugin" CLI subcommand.
+options:
+  name:
+    description:
+      - Name of the plugin to operate on.
+    required: true
+    type: str
+    aliases:
+      - plugin_name
+
+  alias:
+    description:
+      - Alias of the plugin to operate on. Same plugin can be installed with different alias.
+    type: str
+    aliases:
+      - plugin_alias
+
+  plugin_options:
+    description:
+      - Dictionary of plugin settings.
+    type: dict
+
+  state:
+    description:
+      - C(absent) remove the plugin.
+      - C(present) install the plugin, if it does not already exist.
+    default: present
+    choices:
+      - absent
+      - present
+
+extends_documentation_fragment:
+  - docker
+  - docker.docker_py_2_documentation
+
+author:
+  - Vladimir Porshkevich (@porshkevich)
+
+requirements:
+  - "python >= 2.7"
+  - "docker-py >= 2.6.0"
+  - "Please note that the L(docker-py,https://pypi.org/project/docker-py/) Python
+    module has been superseded by L(docker,https://pypi.org/project/docker/)
+    (see L(here,https://github.com/docker/docker-py/issues/1310) for details).
+    Version 2.1.0 or newer is only available with the C(docker) module."
+  - "Docker API >= 1.25"
+'''
+
+EXAMPLES = '''
+- name: Create a plugin
+  docker_plugin:
+    name: plugin_one
+
+- name: Remove a plugin
+  docker_plugin:
+    name: plugin_one
+    state: absent
+
+- name: Create a plugin with options
+  docker_plugin:
+    name: weaveworks/net-plugin:latest_release
+    alias: weave-net-plugin
+    plugin_options:
+      IPALLOC_RANGE: "10.32.0.0/12"
+      WEAVE_PASSWORD: XXXXXXXX
+'''
+
+RETURN = '''
+facts:
+    description: Plugin inspection results for the affected plugin.
+    returned: success
+    type: dict
+    sample: {}
+'''
+
+try:
+    from docker.errors import APIError, NotFound
+    from docker.models.plugins import Plugin
+    from docker import DockerClient
+except ImportError:
+    # missing docker-py handled in ansible.module_utils.docker_common
+    pass
+
+from ansible.module_utils.docker.common import DockerBaseClass, AnsibleDockerClient
+from ansible.module_utils.six import iteritems, text_type
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+        self.client = client
+
+        self.plugin_name = None
+        self.plugin_alias = None
+        self.plugin_options = None
+        self.debug = None
+
+        for key, value in iteritems(client.module.params):
+            setattr(self, key, value)
+
+
+class DockerPluginManager(object):
+
+    def __init__(self, client):
+        self.client = client
+
+        self.dclient = DockerClient()
```

Comment:
I think this should be an extension to ansible.module_utils.docker.common instead of a direct call of a docker library class methods

Reply:
`self.client` is already derived from `DockerClient` (and ready to use), there's no need for `self.dclient`.

Reply:
Sorry, but `self.client` is `docker.APIClient` not `docker.DockerClient`

Reply:
> 
> 
> I think this should be an extension to ansible.module_utils.docker.common instead of a direct call of a docker library class methods

I understand you, but I have no idea what to change in `ansible.module_utils.docker.common` for this

Reply:
Ah, `docker.DockerClient` not our `common.DockerClient`. Sorry, got confused :)

There's one problem with using `docker.DockerClient`: it needs to be set-up similarly to the API client, otherwise it might talk to a completely different docker daemon (or not talk to a docker daemon at all, while the other one does). To set it up the same way, you should add a method to `common.AnsibleDockerClient` which sets it up for you (and call that one to get your instance).

As a first iteration, you can also do the setup in your module (and not touch `common.AnsibleDockerClient`), but you should look at how `common.AnsibleDockerClient` does the setup of the API client. The relevant lines are: https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/docker/common.py#L230-L237
Maybe you can simply pass `client._connect_params` on to `docker.DockerClient`'s constructor and everything is fine. Looking at how `docker.DockerClient` works (https://github.com/docker/docker-py/blob/master/docker/client.py#L39-L40), this seems to be the correct approach.

Reply:
`docker.DockerClient` use `docker.APIClient` as property. `common.AnsibleDockerClient` extend `docker.APIClient`. Instead of setting up a new `docker.APIClient`, I simply replace it not configured with what we already have `common.AnsibleDockerClient` in next line.

Reply:
If this is still not the right decision, I can make a separate option that includes changes to `common.AnsibleDockerClient`.

Reply:
It's probably not a good long-term solution, but currently it's fine, especially because docker-py's `DockerClient` at the moment simply creates `self.api = APIClient(...)` and does nothing else. Once this PR is merged, we can improve this behavior in a bugfix PR (which can also be merged after next Thursday).

Reply:
While working on #53906, I explored the docker client creation process a bit more in detail. You can use `self.dclient = DockerClient(**self.client._connect_params)` to initialize the client with the same arguments as `self.client` is initialized. Relying on a Ansible docker internal API should be safer than relying on docker-py internal API.